src/pam_google_authenticator.c

```diff
@@ -1493,6 +1513,134 @@ static int check_timebased_code(pam_handle_t *pamh, const char*secret_filename,
   return 1;
 }
 
+/* 
+ * Add a 'config' variable that says we logged in from a particular place
+ * at a particular time.  Only remembers the last 10 logins, which 
+ * are replaced in LRU order.
+ *
+ * Returns 0 on success.
+ */
+int
+update_logindetails(pam_handle_t *pamh, const Params *params, char **buf) {
+  const char *rhost = get_rhost(pamh, params);
+  const time_t now = get_time();
+  time_t oldest = now;
+  int oldest_index = -1;
+  unsigned long when = 0;
+  int  found = 0;
+  char *line;
+  char name[] = "LAST ";
+
+  if (rhost == NULL) {
+    return -1;
+  }
+
+  for (int i = 0; i < 10; i++) {
+    char host[40]; /* Max len of ipv6 address is 8*4 digits plus 7 colons.
+                    * Plus trailing NUL is 40
+                    */
+
+    name[4] = i + '0';
+    line = get_cfg_value(pamh, name, *buf);
+    if (line == &oom) {
+      /* Fatal! */
+      return -1;
+    }
+
+    if (!line) {
+      /* Make first empty line the oldest */
+      if (oldest) {
+        oldest_index = i;
+        oldest = 0;
+      }
+      continue;
+    }
+
+    if (sscanf(line, " %40[0-9a-fA-F:.] %lu ", host, &when) != 2) {
+      log_message(LOG_ERR, pamh, "Malformed LAST%d line", i);
+      continue;
+    }
+
+    if (!strcmp(host, rhost)) {
+      found = 1;
+      break;
+    }
+        
+    if (when < oldest) {
+      oldest_index = i;
+      oldest = when;
+    }
+
+    free(line);
+  }
+
+  if (!found) {
+    /* Loop completed all ten iterations */
+    name[4] = oldest_index + '0';
+  } else {
+    free(line);
+  }
+
+  /* 
+   * Max length in decimal digits of a 64 bit number is (64 log 2) + 1 
+   * Plus space and NUL termination, is 23.
+   * Max len of ipv6 address is 39.
+   */
+  char value[64];
+  memset(value, 0, sizeof value);
+
+  snprintf(value, sizeof value, "%s %lu", rhost, (unsigned long)now);
+  set_cfg_value(pamh, name, value, buf);
+
+  return 0;
+}
+
+/*
+ * Return non-zero if the last login from the same host as this one was 
+ * successfully authenticated within the grace period.
+ */
+int
+within_grace_period(pam_handle_t *pamh, const Params *params, char *buf) {
+  const char *rhost = get_rhost(pamh, params);
+  const time_t now = get_time();
+  const time_t grace = params->grace_period;
+  unsigned long when = 0;
+  char match[128];
+  char *line;
+
+  if (rhost == NULL) {
+    return 0;
+  }
+  snprintf(match, sizeof match - 1, " %s %%lu ", rhost);
```

Comment:
```suggestion
  snprintf(match, sizeof match, " %s %%lu ", rhost);
```

All other code in the file assumes `snprintf` that takes whole buffer size. So for consistency here too.

Reply:
Done.